### PR TITLE
DynamicLossScale

### PR DESCRIPTION
**Pull request by liujuncheng**
https://docs.nvidia.com/deeplearning/performance/mixed-precision-training/index.html

> 1. Maintain a primary copy of weights in FP32.
> 2. Initialize S to a large value.
> 3. For each iteration:
>   a. Make an FP16 copy of the weights.
>   b. Forward propagation (FP16 weights and activations).
>   c. Multiply the resulting loss with the scaling factor S.
>   d. Backward propagation (FP16 weights, activations, and their gradients).
>   e. If there is an Inf or NaN in weight gradients:  i. Reduce S.  ii. Skip the weight update and move to the next iteration.
>   f. Multiply the weight gradient with 1/S.
>   g. Complete the weight update (including gradient clipping, etc.).
>   h. If there hasn’t been an Inf or NaN in the last N iterations, increase S.

```python
    loss_scale_policy = flow.optimizer.loss_scale.dynamic_loss_scale(increment_period=20);
    return flow.optimizer.AdamW(lr_scheduler, epsilon=1e-6, weight_decay=args.weight_decay_rate,
                                weight_decay_excludes=["bias", "LayerNorm", "layer_norm"],
                                grad_clipping=flow.optimizer.grad_clipping.by_global_norm(1.0),
                                loss_scale_policy=loss_scale_policy)

```

### CODE REVIEW

File: oneflow/user/kernels/dynamic_loss_scale_schedule_kernel.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
+/*
```
Comment (yuanms2) [inline oneflow/user/kernels/dynamic_loss_scale_schedule_kernel.cpp:1]: 在cpu和gpu上执行loss scale 效果上有什么区别么

File: oneflow/core/job_rewriter/dynamic_loss_scale_schedule_pass.cpp [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  OperatorConf loss_scale_val_op_conf{};
+  const std::string loss_scale_var_lbn = GenLogicalBlobName(
+      loss_scale_var_op_conf.name(), loss_scale_var_op_conf.variable_conf().out());
+  {
+    loss_scale_val_op_conf.set_name(loss_scale_var_op_conf.name() + "-Identity");
+    IdentityOpConf* identity_conf = loss_scale_val_op_conf.mutable_identity_conf();
+    identity_conf->set_in(loss_scale_var_lbn);
+    identity_conf->set_out("out");
+  }
+  // will be replaced by real count of not finite
```
Comment (yuanms2) [inline oneflow/core/job_rewriter/dynamic_loss_scale_schedule_pass.cpp:86]: 这里的意思是没有真实检查nan?

Comment (liujuncheng) [reply oneflow/core/job_rewriter/dynamic_loss_scale_schedule_pass.cpp:86]: > 这里的意思是没有真实检查nan?

loss scale 在后向构图一开始就需要用到，但是 count_not_finite 在后向构图完成后才能得到，这里先插入了一个占位op，在后向构图完成后得到真实的 count_not_finite 后会替换这个 op

File: oneflow/user/kernels/dynamic_loss_scale_schedule_kernel.cpp (hunk 1 above)
Comment (liujuncheng) [reply oneflow/user/kernels/dynamic_loss_scale_schedule_kernel.cpp:1]: > 在cpu和gpu上执行loss scale 效果上有什么区别么

这个op是维护loss scale和good step的，三个输入都是标量，cpu/gpu在性能上不会有差别，目前这个op都是工作在cpu上面的

File: oneflow/core/job_rewriter/autograd.cpp [hunk 3]
```diff
@@ -133,20 +135,65 @@ std::function<bool(const LogicalBlobId&, const std::string&)> MakePredicatorHasD
...
+    const DataType data_type = op_graph.GetLogicalBlobDesc(lbi).data_type();
+    if (data_type == DataType::kFloat) {
+      loss_scale_val_lbn = dynamic_loss_scale_state.loss_scale_val_lbn();
+    } else {
+      auto cast_op =
+          user_op::UserOpConfWrapperBuilder(lbi.op_name() + "_" + lbi.blob_name() + "_grad_Cast")
+              .Op("cast")
+              .Input("in", dynamic_loss_scale_state.loss_scale_val_lbn())
+              .Output("out")
+              .Attr<DataType>("dtype", data_type)
```
[the remark this answers is not in the document]
Comment (liujuncheng) [reply oneflow/core/job_rewriter/autograd.cpp:170]: op_confs里面的op调用者会添加scope id

**Comment by ouyangyu** [comment]
### ResNet50正确性验证
此次验证都是在金山云机器上单机8卡的情况下对真是数据集(111G)进行验证
### 测试环境
- OneFlow版本为dev_pr_dynamic_loss_scale分支上的commit：[https://github.com/Oneflow-Inc/oneflow/commit/3130186198f58b774ce9d85cc14e25a7f9d99965)
- 测试脚本在Benchmark仓库：https://github.com/Oneflow-Inc/OneFlow-Benchmark/pull/159
- 脚本名称：`Classification/cnns/train_fp16.sh`、`Classification/cnns/train_fp32.sh`
- cuDNN: 7.6.5
- NCCL: 2.7.3
- Python：3.7
### 测试结果
#### fp16结果
1. fp16下50Eporch，batch size为192，日志：[dev_pr_dynamic_loss_scale_resnet50_fp16_50E_1_8_192_training.log](http://oneflow-public.oss-cn-beijing.aliyuncs.com/oneflow_test_log/dev_pr_dynamic_loss_scale/dev_pr_dynamic_loss_scale_resnet50_fp16_50E_1_8_192_training.log)
```
train: epoch 48, iter 100, loss: 1.885468, top_1: 0.781061, top_k: 0.923060, samples/s: 7805.132 1606115586.6729984
train: epoch 48, iter 200, loss: 1.853093, top_1: 0.782161, top_k: 0.923105, samples/s: 8021.406 1606115605.8218343
train: epoch 48, iter 300, loss: 1.843663, top_1: 0.782292, top_k: 0.923320, samples/s: 8021.994 1606115624.9691327
train: epoch 48, iter 400, loss: 1.948191, top_1: 0.782539, top_k: 0.923906, samples/s: 7926.400 1606115644.3474011
train: epoch 48, iter 500, loss: 1.891644, top_1: 0.783125, top_k: 0.924648, samples/s: 7945.766 1606115663.6785338
train: epoch 48, iter 600, loss: 1.897842, top_1: 0.781393, top_k: 0.924505, samples/s: 7957.238 1606115682.9817915
train: epoch 48, iter 700, loss: 1.891153, top_1: 0.782435, top_k: 0.923496, samples/s: 7919.255 1606115702.3773646
train: epoch 48, iter 800, loss: 1.850084, top_1: 0.780495, top_k: 0.923255, samples/s: 7932.863 1606115721.7399113
Saving model to ./output/snapshots/model_save-20201123124638/snapshot_epoch_48.
validation: epoch 48, iter 125, top_1: 0.759100, top_k: 0.928120, samples/s: 2824.212 1606115745.7964754
train: epoch 49, iter 100, loss: 1.924866, top_1: 0.782454, top_k: 0.923483, samples/s: 7849.873 1606115765.2162158
train: epoch 49, iter 200, loss: 1.862771, top_1: 0.781602, top_k: 0.923945, samples/s: 8025.308 1606115784.3557727
train: epoch 49, iter 300, loss: 1.867674, top_1: 0.781914, top_k: 0.923353, samples/s: 8025.905 1606115803.4936998
train: epoch 49, iter 400, loss: 1.916649, top_1: 0.780723, top_k: 0.923197, samples/s: 7931.308 1606115822.8600266
train: epoch 49, iter 500, loss: 1.930456, top_1: 0.782109, top_k: 0.922461, samples/s: 7905.935 1606115842.2884638
train: epoch 49, iter 600, loss: 1.923271, top_1: 0.781328, top_k: 0.923236, samples/s: 7966.708 1606115861.5687034
train: epoch 49, iter 700, loss: 1.844358, top_1: 0.783809, top_k: 0.923854, samples/s: 7909.132 1606115880.9892428
train: epoch 49, iter 800, loss: 1.851614, top_1: 0.782585, top_k: 0.924642, samples/s: 7940.560 1606115900.3330126
Saving model to ./output/snapshots/model_save-20201123124638/snapshot_epoch_49.
validation: epoch 49, iter 125, top_1: 0.758760, top_k: 0.927640, samples/s: 2890.491 1606115924.0109794
```
### DLperf测试
此次验证都是在类脑机器上单机单卡、单机8卡、2机8卡的情况下对真是数据集(111G)进行测速
- OneFlow版本为dev_pr_dynamic_loss_scale分支上的commit：[https://github.com/Oneflow-Inc/oneflow/commit/3130186198f58b774ce9d85cc14e25a7f9d99965)
- 测试脚本在Benchmark仓库：https://github.com/Oneflow-Inc/OneFlow-Benchmark/pull/159
#### ResNet50
OneFlow 0.2版本DLPerf结果：https://github.com/Oneflow-Inc/DLPerf/blob/master/OneFlow/ConvNets/resnet50_oneflow_v0.2_report_1009.md
##### FP16
| num_nodes  | gpu_num_per_node  | batch_size_per_device  | throughput | speedup   |
| ------------ | -------------------- | ----------------------- | ----------- | ---------- |
| 1                    |          1                        | 256                                 | 1441.36 | 1.00 |
| 1 | 8 | 256 | 10303.63 | 7.15 |
| 2 | 8 | 256 | 17397.58 | 12.07 |
日志：[resnet50_fp16_b256_logs.zip](http://oneflow-public.oss-cn-beijing.aliyuncs.com/oneflow_test_log/dev_pr_dynamic_loss_scale/resnet50_fp16_b256_logs.zip)

**Comment by ouyangyu** [comment]
### BERT性能
#FP32 with clip
| num_nodes | gpu_num_per_node | batch_size_per_device | throughput | speedup |
| -------- | -------- | -------- | -------- | -------- |
| 1 | 1 | 64 | 157.07 | 1.00 |
| 1 | 8 | 64 | 1205.37 | 7.67 |
| 2 | 8 | 64 | 2323.50 | 14.79 |

日志：[bert_fp32_b64_logs.zip](http://oneflow-public.oss-cn-beijing.aliyuncs.com/oneflow_test_log/dev_pr_dynamic_loss_scale/bert_fp32_b64_logs.zip)

**Comment by ouyangyu** [comment]
#### ResNet50正确性
fp16下90Eporch，batch size为192，

日志：[dev_pr_dynamic_loss_scale_resnet50_fp16_50E_1_8_192_training.log](http://oneflow-public.oss-cn-beijing.aliyuncs.com/oneflow_test_log/dev_pr_dynamic_loss_scale/dev_pr_dynamic_loss_scale_resnet50_fp16_50E_1_8_192_training.log)、[dev_pr_dynamic_loss_scale_resnet50_fp16_90E_1_8_192_oneflow.INFO](http://oneflow-public.oss-cn-beijing.aliyuncs.com/oneflow_test_log/dev_pr_dynamic_loss_scale/dev_pr_dynamic_loss_scale_resnet50_fp16_90E_1_8_192_oneflow.INFO)
```
Saving model to ./output/snapshots/model_save-20201123160924/snapshot_epoch_88.
validation: epoch 88, iter 125, top_1: 0.770380, top_k: 0.932920, samples/s: 2896.833 1606135151.546864
train: epoch 89, iter 100, loss: 1.787400, top_1: 0.815599, top_k: 0.938366, samples/s: 7673.906 1606135171.3983235
train: epoch 89, iter 200, loss: 1.815307, top_1: 0.815658, top_k: 0.937786, samples/s: 7974.478 1606135190.6598067
train: epoch 89, iter 300, loss: 1.733322, top_1: 0.816914, top_k: 0.938438, samples/s: 7946.067 1606135209.9900386
train: epoch 89, iter 400, loss: 1.768651, top_1: 0.816432, top_k: 0.938340, samples/s: 7884.779 1606135229.4706671
train: epoch 89, iter 500, loss: 1.749225, top_1: 0.816536, top_k: 0.937949, samples/s: 7879.342 1606135248.964715
train: epoch 89, iter 600, loss: 1.751591, top_1: 0.816504, top_k: 0.938861, samples/s: 7871.351 1606135268.478424
train: epoch 89, iter 700, loss: 1.772364, top_1: 0.817435, top_k: 0.939036, samples/s: 7907.347 1606135287.903488
train: epoch 89, iter 800, loss: 1.791631, top_1: 0.817415, top_k: 0.939232, samples/s: 7849.873 1606135307.4706829
Saving model to ./output/snapshots/model_save-20201123160924/snapshot_epoch_89.
validation: epoch 89, iter 125, top_1: 0.770160, top_k: 0.933060, samples/s: 2887.330 1606135331.1946614
```

**Comment by ouyangyu** [comment]
#### ResNet50正确性
fp16下50Eporch，batch size为192，
commit : 3bf8402
日志：[last_resnet50_fp16_50E_1_8_192_training.log](http://oneflow-public.oss-cn-beijing.aliyuncs.com/oneflow_test_log/dev_pr_dynamic_loss_scale/last_resnet50_fp16_50E_1_8_192_training.log)、[last_resnet50_fp16_50E_1_8_192_oneflow.INFO](http://oneflow-public.oss-cn-beijing.aliyuncs.com/oneflow_test_log/dev_pr_dynamic_loss_scale/last_resnet50_fp16_50E_1_8_192_oneflow.INFO)
```
train: epoch 48, iter 100, loss: 1.838229, top_1: 0.781621, top_k: 0.923776, samples/s: 9593.354 1606197404.8393943
train: epoch 48, iter 200, loss: 1.858704, top_1: 0.781927, top_k: 0.923288, samples/s: 9901.942 1606197420.35167
train: epoch 48, iter 300, loss: 1.841751, top_1: 0.783177, top_k: 0.924440, samples/s: 9876.140 1606197435.9046032
train: epoch 48, iter 400, loss: 1.894127, top_1: 0.782480, top_k: 0.923346, samples/s: 9764.474 1606197451.634571
train: epoch 48, iter 500, loss: 1.863208, top_1: 0.781478, top_k: 0.923685, samples/s: 9748.521 1606197467.3908794
train: epoch 48, iter 600, loss: 1.862235, top_1: 0.781296, top_k: 0.924284, samples/s: 9699.159 1606197483.2272735
train: epoch 48, iter 700, loss: 1.905760, top_1: 0.781426, top_k: 0.923177, samples/s: 9808.559 1606197498.8870442
train: epoch 48, iter 800, loss: 1.828721, top_1: 0.781191, top_k: 0.923021, samples/s: 9819.643 1606197514.529275
Saving model to ./output/snapshots/model_save-20201124114958/snapshot_epoch_48.
validation: epoch 48, iter 125, top_1: 0.758280, top_k: 0.928420, samples/s: 2116.776 1606197543.51824
train: epoch 49, iter 100, loss: 1.860581, top_1: 0.782500, top_k: 0.923952, samples/s: 9604.018 1606197559.339539
train: epoch 49, iter 200, loss: 1.915841, top_1: 0.781510, top_k: 0.923639, samples/s: 9888.300 1606197574.872934
train: epoch 49, iter 300, loss: 1.983146, top_1: 0.782643, top_k: 0.922943, samples/s: 9600.721 1606197590.8717315
train: epoch 49, iter 400, loss: 1.909353, top_1: 0.781803, top_k: 0.922565, samples/s: 9692.438 1606197606.7190819
train: epoch 49, iter 500, loss: 1.895887, top_1: 0.781224, top_k: 0.923333, samples/s: 9581.075 1606197622.7508307
train: epoch 49, iter 600, loss: 1.854682, top_1: 0.782832, top_k: 0.923268, samples/s: 9841.889 1606197638.3574622
train: epoch 49, iter 700, loss: 1.919612, top_1: 0.780755, top_k: 0.923561, samples/s: 9648.848 1606197654.276487
train: epoch 49, iter 800, loss: 1.843291, top_1: 0.780898, top_k: 0.923613, samples/s: 9887.736 1606197669.8108966
Saving model to ./output/snapshots/model_save-20201124114958/snapshot_epoch_49.
validation: epoch 49, iter 125, top_1: 0.758620, top_k: 0.928900, samples/s: 2125.331 1606197698.431872
```

**Review by chengtbf — APPROVED** [review]

**Review by lixinqi — APPROVED** [review]
